pkg/csi/service/wcp/controller.go

```diff
@@ -519,10 +520,34 @@ func (c *controller) createBlockVolume(ctx context.Context, req *csi.CreateVolum
 	}
 
 	// Calculate accessible topology for the provisioned volume in case of topology aware environment.
-	if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.TKGsHA) {
+	if isTKGSHAEnabled {
 		if zoneLabelPresent && !hostnameLabelPresent {
-			// Calculate accessible topology for the provisioned volume.
 			selectedDatastore := volumeInfo.DatastoreURL
+			// CreateBlockVolumeUtil with idempotency enabled does not return datastore
+			// information when it uses the cached information from CR. In such cases,
+			// querying the volume to retrieve the datastore URL.
+			if selectedDatastore == "" {
+				queryFilter := cnstypes.CnsQueryFilter{
+					VolumeIds: []cnstypes.CnsVolumeId{
+						{
+							Id: volumeInfo.VolumeID.Id,
+						},
+					},
+				}
+				querySelection := cnstypes.CnsQuerySelection{
+					Names: []string{utils.CnsQuerySelectionName_DATASTORE_URL},
+				}
+				queryResult, err := utils.QueryVolumeUtil(ctx, c.manager.VolumeManager, queryFilter, &querySelection,
+					commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.AsyncQueryVolume))
```

Comment:
regarding `async-query-volume` flag, can we not set this to `true`?
as this is already released before `tkgs-ha` feature?

Reply:
I thought of doing that but decided to keep it this way and remove it when we remove the FSS itself. Is that okay?

Reply:
It impacts the performance of create volume operation and we are not going to disable `async-query-volume` so we are good.

Reply:
Okay changed